tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/connect/ValidateAMQPErrorsTest.java

```diff
@@ -0,0 +1,643 @@
+      server.start();
+
+      Wait.assertTrue(() -> AssertionLoggerHandler.findText("AMQ111001"));
+      Assert.assertEquals(6, AssertionLoggerHandler.countText("AMQ119018")); // 0..5 = 6
+   }
+
+   /**
+    * Refuse the first mirror link, verify broker handles it and reconnects
+    *
+    * @throws Exception
+    */
+   @Test
+   public void testReconnectAfterMirrorLinkRefusal() throws Exception {
+      startVerx();
+
+      AtomicInteger errors = new AtomicInteger(0);
+
+      AtomicInteger messagesReceived = new AtomicInteger(0);
+
+      List<ProtonConnection> connections = Collections.synchronizedList(new ArrayList<ProtonConnection>());
+      List<ProtonConnection> disconnected = Collections.synchronizedList(new ArrayList<ProtonConnection>());
+      AtomicInteger refusedLinkMessageCount = new AtomicInteger();
+      AtomicInteger linkOpens = new AtomicInteger(0);
+
+      mockServer = new MockServer(vertx, serverConnection -> {
+         serverConnection.disconnectHandler(c -> {
+            disconnected.add(serverConnection);
+         });
+
+         serverConnection.openHandler(c -> {
+            connections.add(serverConnection);
+            serverConnection.open();
+         });
+
+         serverConnection.closeHandler(c -> {
+            serverConnection.close();
+            connections.remove(serverConnection);
+         });
+
+         serverConnection.sessionOpenHandler(session -> {
+            session.open();
+         });
+
+         serverConnection.receiverOpenHandler(serverReceiver -> {
+            Target remoteTarget = serverReceiver.getRemoteTarget();
+            String remoteAddress = remoteTarget == null ? null : remoteTarget.getAddress();
+            if (remoteAddress == null || !remoteAddress.startsWith(ProtonProtocolManager.MIRROR_ADDRESS)) {
+               errors.incrementAndGet();
+               System.err.println("Receiving address as " + remoteAddress);
+               return;
+            }
+            if (linkOpens.incrementAndGet() != 2) {
+               logger.debug("Link Opens::" + linkOpens);
+               logger.debug("ServerReceiver = " + serverReceiver.getTarget());
+               serverReceiver.setTarget(null);
+
+               serverReceiver.handler((del, msg) -> {
+                  refusedLinkMessageCount.incrementAndGet();
+                  logger.debug("Should not have got message on refused link: " + msg);
+               });
+
+               //      Adding exagerated delay between 'refusing attach' and following detach, and
+               //      didnt set prefetch to 0, so credit will be given during explicit open() (credit wouldnt be given if we
+               //      removed this and just let the close() cause the needed attach frame before sending the detach frame).
+               //      If removing, get rid of refusedLinkMessageCount and related usage since it then cant happen.
+               serverReceiver.open();
+
+               vertx.setTimer(5000, x -> {
```

Comment:
```suggestion
               vertx.setTimer(20, x -> {
```

I hugely exaggerated this originally to make the issues visible to the eye in the 'debug test tool'. There was also a comment at the time saying to reduce it to a small-but-delivery-feasible value if keeping. Say 20ms or something.